src/library/features/crates/cratehierarchy.cpp

```diff
@@ -0,0 +1,577 @@
+    FwdSqlQuery query(
+      m_database, QString(
+        "INSERT INTO %1(%2, %3, %4) "
+        "SELECT p.%2, c.%3, p.%4 + c.%4 + 1 "
+        "FROM %1 p, %1 c "
+        "WHERE p.%3 = :parent AND c.%2 = :child").arg(
+          CRATE_CLOSURE_TABLE,
+          CLOSURE_PARENTID,
+          CLOSURE_CHILDID,
+          CLOSURE_DEPTH));
+
+    if (!query.isPrepared()) {
+        return false;
+    }
+
+    query.bindValue(":parent", parent);
+    query.bindValue(":child", child);
+
+    if (!query.execPrepared()) {
+        return false;
+    }
+
+    return true;
+}
+
+void CrateHierarchy::deleteCrate(CrateId id) const {
+    {
+        FwdSqlQuery query(
+          m_database, QString(
+            "DELETE FROM cratePath WHERE crateId = :id"));
+        query.bindValue(":id", id);
+        if (!query.isPrepared()) {
+            return;
+        }
+        if (!query.execPrepared()) {
+            return;
+        }
+    }
+
+    {
+        FwdSqlQuery query(
+          m_database, QString(
+            "DELETE FROM crateClosure WHERE childId = :id"));
+        query.bindValue(":id", id);
+        if (!query.isPrepared()) {
+            return;
+        }
+        if (!query.execPrepared()) {
+            return;
+        }
+    }
+}
+
+QStringList CrateHierarchy::collectCrateIdsByCrateNameLike(const QString& crateNameLike) const {
+    FieldEscaper escaper(m_database);
+    QString escapedArgument = escaper.escapeString(kSqlLikeMatchAll + crateNameLike + kSqlLikeMatchAll);
+
+    FwdSqlQuery query(
+      m_database, QString(
+        "SELECT DISTINCT %1 FROM %2 "
+        "WHERE %3 LIKE %4").arg(
+          CRATETABLE_ID,
+          CRATE_TABLE,
+          CRATETABLE_NAME,
+          escapedArgument));
+
+    QStringList ids;
+    if (query.execPrepared())
+        while (query.next()) {
+            ids << query.fieldValue(0).toString();
+    }
+
+}
+
+QString CrateHierarchy::formatQueryForTrackIdsByCrateNameLikeRecursively(const QString& crateNameLike) const {
+    QStringList childIds;
+    QStringList ids = collectCrateIdsByCrateNameLike(crateNameLike);
+    
+    for (const auto& id : ids) {
+        childIds = childIds + collectChildCrateIds(CrateId(id.toInt()));
+    }
+
+    return QString(
+        "SELECT DISTINCT %1 FROM %2 "
+        "JOIN %3 ON %4=%5 "
+        "WHERE %4 IN (%6) "
+        "OR %4 IN (%7) "
+        "ORDER BY %1").arg(
+            CRATETRACKSTABLE_TRACKID,
+            CRATE_TRACKS_TABLE,
+            CRATE_PATH_TABLE,
+            PATHTABLE_CRATEID,
+            CRATETRACKSTABLE_CRATEID,
+            ids.join(","),
+            childIds.join(","));
+}
+
+bool CrateHierarchy::isNameValidForHierarchy(const QString& newName,
+                                             const Crate& selectedCrate,
+                                             const Crate& parent) const {
+    // if crate is valid it means we are trying to rename an existing crate
```

Comment:
Please remember that users will expect a file-system-like behavior of the create hierarchy. How do you want to explain those complicated naming rules to users if this function returns false? How could a users know why a renaming failed?

I again propose to only check for uniqueness of names among siblings and not upwards and downwards within the hierarchy.